nucliadb/nucliadb/ingest/orm/processor.py

```diff
@@ -372,64 +371,38 @@ async def txn(
         try:
             for message in messages:
+                reindex = reindex or message.reindex
                 if resource is not None:
                     assert resource.uuid == message.uuid
-                result = await self.apply_resource(message, kb, resource)
-
-                if result is None:
+                try:
+                    resource, _created = await self.apply_message(message, kb, resource)
+                    created = created or _created
+                except ApplyMessageError as exc:
+                    logger.warning(str(exc))
                     continue
 
-                resource, _created = result
-                created = created or _created
-
-            if resource:
-                await resource.compute_global_text()
-                await resource.compute_global_tags(resource.indexer)
-                if message.reindex:
-                    # when reindexing, let's just generate full new index message
-                    resource.replace_indexer(await resource.generate_index_message())
-
-            if resource and resource.modified:
-                shard_id = await kb.get_resource_shard_id(uuid)
-                node_klass = get_node_klass()
-
-                if shard_id is not None:
-                    shard = await kb.get_resource_shard(shard_id, node_klass)
-
-                if shard is None:
-                    # It's a new resource, get current active shard to place
-                    # new resource on
-                    shard = await node_klass.get_current_active_shard(txn, kbid)
-                    if shard is None:
-                        # no shard available, create a new one
-                        similarity = await kb.get_similarity()
-                        shard = await node_klass.create_shard_by_kbid(
-                            txn, kbid, similarity=similarity
-                        )
-                    await kb.set_resource_shard_id(uuid, shard.sharduuid)
-
-                if shard is not None:
-                    counter = await shard.add_resource(
-                        resource.indexer.brain, seqid, partition=partition, kb=kbid
-                    )
-                    if (
-                        counter is not None
-                        and counter.fields > settings.max_shard_fields
-                    ):
-                        # shard is full, create a new one so next resource
-                        # is placed on a new shard
-                        similarity = await kb.get_similarity()
-                        shard = await node_klass.create_shard_by_kbid(
-                            txn, kbid, similarity=similarity
-                        )
-                else:
-                    raise AttributeError("Shard is not available")
+            if resource is None:
+                raise NoResourceError()
+
+            # TODO: in case of reindex = True, shouldn't we do that after generate_index_message?
```

Comment:
I think that should be right? I guess it was done this way to ensure we didn't miss anything.